static/js/services/message-list-utils.js

```diff
@@ -0,0 +1,45 @@
+var _ = require('underscore');
+
+angular.module('inboxServices').factory('MessageListUtils',
+  function(
+  ) {
+    'use strict';
+    'ngInject';
+
+    var removeDeleted = function(allMessages, changedMessages) {
+      allMessages.forEach(function(message, i, scoped) {
+        if (!_.some(changedMessages, function(updated) {
+          return message.key === updated.key;
+        })) {
+          scoped.splice(i, 1);
+        }
+      });
+    };
```

Comment:
I don't think this is correct. If you splice out of the array you're iterating over aren't you going to skip some elements? That's why the initial implementation iterated in reverse so the splicing wouldn't affect the index of elements not yet iterated over. Try writing a unit test with two elements in allMessages both of which should be deleted and see if they both are.

Reply:
You're right, my test wasn't good enough! To me the forEach docs implied that it would deal with it correctly, but apparently not. I'll probably revert to the old code